File: workloads/resnet/resnet_train.py
print("resnet_train")

import torch
import torchvision.models as models
import torch.nn as nn
import torch.optim as optim
import time

print("starting up")

device = torch.device("cpu")
model = models.resnet50().to(device)

# Get the current number of threads (optional)
# current_threads = torch.get_num_threads()
# print(f"Current number of threads: {current_threads}")

# # Set the desired number of threads
# desired_threads = 4  # Example: set to 4 threads
# torch.set_num_threads(desired_threads)

# # Verify the new number of threads (optional)
# new_threads = torch.get_num_threads()
# print(f"New number of threads: {new_threads}")

# Synthetic data
batch_size = 512
dummy_input = torch.randn(batch_size, 3, 224, 224).to(device)
dummy_target = torch.randint(0, 1000, (batch_size,)).to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Config
epochs = 1
iters_per_epoch = 1

model.train()
for epoch in range(1, epochs + 1):
    start = time.time()
    for i in range(iters_per_epoch):
        print(f"Iteration {i + 1}/{iters_per_epoch}")
        optimizer.zero_grad()
        output = model(dummy_input)
        loss = criterion(output, dummy_target)
        loss.backward()
        optimizer.step()
    end = time.time()
    images_per_sec = (iters_per_epoch * batch_size) / (end - start)
    print(f"Epoch {epoch}: {images_per_sec:.2f} images/sec")

File: workloads/resnet/resnet_train.cc
#include <torch/torch.h>
#include <torch/script.h>

#include <iostream>
#include <vector>
#include <chrono>
#include <thread>

using clock_type = std::chrono::high_resolution_clock;

int main() {
    std::cout << "ResNet50 TorchScript benchmark" << std::endl;

    // ---------------- Configuration ----------------
    const int batch_size = 128;
    const int iters_per_epoch = 2;
    const int epochs = 2;

    torch::Device device(torch::kCPU);

    // Thread limits (important for allocator stability)
    torch::set_num_threads(1);
    torch::set_num_interop_threads(1);

    // ---------------- Load model ----------------
    torch::jit::Module model;
    try {
        model = torch::jit::load("/proj/TppPlus/tpp/libnuma_pgmig/workloads/resnet/build/resnet50.pt");
    } catch (const c10::Error& e) {
        std::cerr << "Error loading resnet50.pt\n";
        return -1;
    }

    model.to(device);
    model.train();

    // ---------------- Persistent input & target ----------------
    auto input = torch::randn(
        {batch_size, 3, 224, 224},
        torch::TensorOptions().device(device)
    );

    auto target = torch::randint(
        0, 1000,
        {batch_size},
        torch::TensorOptions().device(device).dtype(torch::kLong)
    );

    // ---------------- Loss ----------------
    auto criterion = torch::nn::CrossEntropyLoss();

    // ---------------- Optimizer (FIXED) ----------------
    std::vector<torch::Tensor> params;
    for (const auto& p : model.parameters()) {
        params.push_back(p);
    }

    torch::optim::SGD optimizer(
        params,
        torch::optim::SGDOptions(0.01)
    );

    // ---------------- Warm-up (allocate once) ----------------
    {
        std::vector<torch::jit::IValue> inputs;
        inputs.push_back(input);

        auto output = model.forward(inputs).toTensor();
        auto loss = criterion(output, target);

        // FIXED backward call
        loss.backward(/*gradient=*/{}, /*retain_graph=*/true);
    }

    // ---------------- Training loop (steady-state) ----------------
    for (int epoch = 1; epoch <= epochs; ++epoch) {
        auto start = clock_type::now();

        for (int i = 0; i < iters_per_epoch; ++i) {
            std::cout << "Iteration "
                      << (i + 1) << "/" << iters_per_epoch << std::endl;

            optimizer.zero_grad(/*set_to_none=*/false);

            std::vector<torch::jit::IValue> inputs;
            inputs.push_back(input);

            auto output = model.forward(inputs).toTensor();
            auto loss = criterion(output, target);

            // KEEP GRAPH ALIVE
            loss.backward(/*gradient=*/{}, /*retain_graph=*/true);
            optimizer.step();
        }

        auto end = clock_type::now();
        double seconds =
            std::chrono::duration<double>(end - start).count();

        double images_per_sec =
            (iters_per_epoch * batch_size) / seconds;

        std::cout << "Epoch " << epoch
                  << ": " << images_per_sec
                  << " images/sec" << std::endl;
    }

    std::cout << "Done. Sleeping..." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(10));

    return 0;
}
